src/coreclr/vm/gchelpers.cpp

```diff
@@ -447,14 +447,24 @@ OBJECTREF AllocateSzArray(MethodTable* pArrayMT, INT32 cElements, GC_ALLOC_FLAGS
             orArray = (ArrayBase*)Alloc(totalSize + MIN_OBJECT_SIZE, flags);
 
             Object* orDummyObject;
-            if ((size_t)orArray % sizeof(double))
+            if (((size_t)orArray % sizeof(double)) != 0)
             {
                 orDummyObject = orArray;
                 orArray = (ArrayBase*)((size_t)orArray + MIN_OBJECT_SIZE);
+                if (flags & GC_ALLOC_ZEROING_OPTIONAL)
```

Comment:
This will likely fix the immediate problem, but I wonder if it should be `gc.cpp` that is responsible for making sure that the sync block is cleared.

If we end up fixing like this, we should update NativeAOT as well?

Reply:
> I wonder if it should be gc.cpp that is responsible for making sure that the sync block is cleared.

This logic is for opportunistic alignment of `double[]` arrays on x86. If you would like to move this logic to gc.cpp (including the syncblock clearing), I would be fine with it. Feel free to do it in a follow up PR - it may be a breaking change for GC/EE interface.

This is x86 specific perf optimization. This logic does not exist in native AOT (native AOT does not support x86 today) so there is nothing to update there.

Reply:
> This logic is for opportunistic alignment of double[] arrays on x86.

Aha, I missed that, this branch is perfectly reasonable, but what about the second branch?

Reply:
The second branch is to zero out the trailing dummy object, which is also reasonable.